packages/grafana-runtime/src/utils/DataSourceWithBackend.ts

```diff
@@ -257,6 +257,19 @@ class DataSourceWithBackend<
   }
 }
 
+function getUrlForRequest<TQuery extends DataQuery = DataQuery>(request: DataQueryRequest<TQuery>): string {
+  if (!config.featureToggles.validatedQueries) {
+    return '/api/ds/query';
+  }
+
+  if (!request.userCanEditDashboard && request.dashboardUid != null && request.panelId != null) {
+    // add additional params to query so we can validate on the backend.
+    return `/api/dashboards/org/${config.bootData.user.orgId}/uid/${request.dashboardUid}/panels/${request.panelId}/query`;
```

Comment:
What's the story around including `orgId` in the URL? Grafana UI always provide the `X-Grafana-Org-Id` header to requests (I think).

Regarding `panelId`, wonder if repeated panels would be tricky to support? Something you've already discussed? In general, the `panelId` is numeric and maybe not that stable? Grafana UI generates it, and with repeated panels feels like it changes, maybe every time you load such a dashboard, but not sure.  Maybe should consider introduction of a panel UID, no idea though what that would look like. Probably would be helpful for those creating dashboards outside of Grafana using "as code" to provide a unique identifier string to each panel rather than a numeric identifier. Something to think about, but probably out of scope for this.

Reply:
In the case of anonymous access the `X-Grafana-Org-Id` header either won't be available or won't be reliable and is required in order to lookup the dashboard on the API side.

Reply:
For repeating panels, I think the best course of action is to say it's not supported for now, but we have added it to the roadmap for validated queries.

Reply:
> In the case of anonymous access the X-Grafana-Org-Id header either won't be available or won't be reliable and is required in order to lookup the dashboard on the API side.

Hmm wasn't aware of that. Then the next question, how is `orgId` set in the browser URL when browsing one of these dashboards?

Reply:
We will most likely use a non-sequential guid that can be decoded or with a lookup in the backend

Reply:
Not sure this API follows are existing conventions well, 

* we have no other API that includes orgId
* This API feels two specific to dashboards, we will need to validate queries for other things in the future. So having a more generic API that can accept more than dashboardUid and panelId in the future would be useful. would expect something like `/api/ds/query-validated`


We also need to pass current template variables (at least in future iteration), so I think this needs to be a POST request. we cannot only change the URL and keep the same request flow as the normal nonvalidated requests.

Reply:
> In the case of anonymous access the X-Grafana-Org-Id header either won't be available or won't be reliable and is required in order to lookup the dashboard on the API side.

This is not true, there can only be one anonymous org so the orgId is always known, even for anonymous orgs. To support multiple anonymous orgs at the same time will require a bigger change to both view URLs and and probably API URLs, but let's leave that for another change.